src/main/java/com/google/firebase/auth/SamlProviderConfig.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.auth;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.api.client.util.Key;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.firebase.auth.ProviderConfig.AbstractCreateRequest;
+import com.google.firebase.auth.ProviderConfig.AbstractUpdateRequest;
+import com.google.firebase.auth.internal.SamlProviderConfigResponse.IdpCertificate;
+import com.google.firebase.auth.internal.SamlProviderConfigResponse.IdpConfig;
+import com.google.firebase.auth.internal.SamlProviderConfigResponse.SpConfig;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Contains metadata associated with a SAML Auth provider.
+ *
+ * <p>Instances of this class are immutable and thread safe.
+ */
+public final class SamlProviderConfig extends ProviderConfig {
+
+  @Key("idpConfig")
```

Comment:
~~Let's move these into the new `SamlProviderConfigResponse` class as well, and handle all deserialization business there. I think this class can just wrap an instance of the response.~~

**EDIT:** I suppose we can't do that since `ProviderConfig` already contains some deserialization logic (providerId, displayName etc). In that case let's define separate top-level classes in the internal package as follows:

* SamlProviderConfigResponse.IdpConfig --> SamlIdpConfig
* SamlProviderConfigResponse.SpConfig --> SamlSpConfig
* SamlProviderConfigResponse.IdpCertificate --> SamlIdpConfig.IdpCertificate

Btw, the option of parsing into maps, and handling the keys manually doesn't sound too onerous in this case either if you want to consider that option:

```
@Key("idpConfig")
private GenericJson idpConfig;

@Key("spConfig")
private GenericJson spConfig;

public String getIdpEntityId() {
    return (String) idpConfig.get("idpEntityId");
}
```

Looks a bit cleaner than adding a bunch of new internal types.

Reply:
The GenericJson approach ended up working well! I thought it was going to be more complicated than it was, since we have nested types.